File: backend/services/user_service.py
"""User profile + role management service."""
from __future__ import annotations
from datetime import datetime, timezone
from bson import ObjectId
from fastapi import HTTPException

from database import get_db
from models.user import User

VALID_ROLES = {"customer", "vendor", "creator"}  # admin only via seed


async def get_user_by_id(user_id: str) -> User | None:
    db = get_db()
    doc = await db.users.find_one({"_id": ObjectId(user_id), "is_deleted": {"$ne": True}})
    return User.from_mongo(doc)


async def update_profile(user_id: str, updates: dict) -> User:
    db = get_db()
    allowed = {"name", "email", "profile_pic", "gender", "dob", "current_role", "city"}
    clean = {k: v for k, v in updates.items() if k in allowed and v is not None}
    if not clean:
        raise HTTPException(400, "No updatable fields provided")

    if "current_role" in clean:
        user = await get_user_by_id(user_id)
        if not user or clean["current_role"] not in user.roles:
            raise HTTPException(400, "current_role must be one of the user's roles")

    clean["updated_at"] = datetime.now(timezone.utc).isoformat()
    await db.users.update_one({"_id": ObjectId(user_id)}, {"$set": clean})
    doc = await db.users.find_one({"_id": ObjectId(user_id)})
    return User.from_mongo(doc)


async def switch_role(user_id: str, role: str) -> User:
    if role not in VALID_ROLES and role != "admin":
        raise HTTPException(400, f"Invalid role: {role}")
    user = await get_user_by_id(user_id)
    if not user:
        raise HTTPException(404, "User not found")
    if role not in user.roles:
        raise HTTPException(400, "You don't have this role. Add it first.")
    db = get_db()
    await db.users.update_one(
        {"_id": ObjectId(user_id)},
        {"$set": {"current_role": role, "updated_at": datetime.now(timezone.utc).isoformat()}},
    )
    doc = await db.users.find_one({"_id": ObjectId(user_id)})
    return User.from_mongo(doc)


async def add_role(user_id: str, role: str) -> User:
    if role not in VALID_ROLES:
        raise HTTPException(400, f"Invalid role: {role}. Allowed: {sorted(VALID_ROLES)}")
    user = await get_user_by_id(user_id)
    if not user:
        raise HTTPException(404, "User not found")
    if role in user.roles:
        return user
    db = get_db()
    await db.users.update_one(
        {"_id": ObjectId(user_id)},
        {
            "$addToSet": {"roles": role},
            "$set": {"updated_at": datetime.now(timezone.utc).isoformat()},
        },
    )
    doc = await db.users.find_one({"_id": ObjectId(user_id)})
    return User.from_mongo(doc)


def serialize(user: User) -> dict:
    return {
        "id": user.id,
        "phone": user.phone,
        "name": user.name,
        "email": user.email,
        "roles": user.roles,
        "current_role": user.current_role,
        "kyc_status": user.kyc_status,
        "profile_pic": user.profile_pic,
        "gender": user.gender,
        "dob": user.dob,
        "is_active": user.is_active,
        # Phase 4a additions
        "is_subscribed_verified": bool(user.is_subscribed_verified),
        "verified_badge": bool(user.is_subscribed_verified) and user.kyc_status == "approved",
        "rating_avg": user.rating_avg,
        "rating_count": user.rating_count,
        "trust_score": user.trust_score,
        "city": user.city,
        # Phase 5 additions
        "referral_code": getattr(user, "referral_code", None),
        "avg_response_time_seconds": getattr(user, "avg_response_time_seconds", None),
        "chat_response_rate": getattr(user, "chat_response_rate", 0.0),
        "created_at": user.created_at,
        "updated_at": user.updated_at,
    }

File: backend/src/services/user.service.js
const User = require('../models/User');
const ApiError = require('../utils/ApiError');

const getUserById = async (id) => {
  const user = await User.findById(id);
  if (!user || user.is_deleted) return null;
  return user;
};

const addRole = async (userId, role) => {
  const validRoles = ['customer', 'vendor', 'creator'];
  if (!validRoles.includes(role)) throw ApiError.badRequest('Invalid role');
  const user = await User.findById(userId);
  if (!user) throw ApiError.notFound('User not found');
  if (user.roles.includes('admin')) {
    throw ApiError.forbidden('Admin accounts cannot add user roles');
  }
  if (!user.roles.includes(role)) {
    user.roles.push(role);
  }
  user.current_role = role;
  user.activeRole = role;
  await user.save();
  try {
    const cache = require('../utils/cache');
    await cache.deleteCache(`user:auth:${userId}`);
  } catch (err) {}
  return user;
};

const switchRole = async (userId, role) => {
  const user = await User.findById(userId);
  if (!user) throw ApiError.notFound('User not found');
  if (user.roles.includes('admin') && role !== 'admin') {
    throw ApiError.forbidden('Admin accounts cannot switch to non-admin roles');
  }
  if (!['customer', 'vendor', 'creator', 'admin'].includes(role)) {
    throw ApiError.badRequest(`Invalid target role: ${role}`);
  }
  if (!user.roles.includes(role)) {
    user.roles.push(role);
  }
  user.current_role = role;
  user.activeRole = role;
  await user.save();
  try {
    const cache = require('../utils/cache');
    await cache.deleteCache(`user:auth:${userId}`);
  } catch (err) {}

  let isOnboardingRequired = false;
  let targetOnboardingPath = null;
  let targetDashboardPath = null;

  if (role === 'vendor') {
    targetDashboardPath = '/vendor/dashboard';
    const vp = user.vendorProfile || {};
    const isComplete = Boolean(vp.shopName || vp.businessName || vp.store_name);
    if (!isComplete) {
      isOnboardingRequired = true;
      targetOnboardingPath = '/vendor/onboarding';
    }
  } else if (role === 'creator') {
    targetDashboardPath = '/creator/dashboard';
    const cp = user.creatorProfile || {};
    const isComplete = Boolean(cp.displayName || cp.name);
    if (!isComplete) {
      isOnboardingRequired = true;
      targetOnboardingPath = '/creator/onboarding';
    }
  } else if (role === 'customer') {
    targetDashboardPath = '/customer/home';
    const custp = user.customerProfile || {};
    const isComplete = Boolean(
      custp.interestsSelectedAt || (Array.isArray(custp.interests) && custp.interests.length >= 5)
    );
    if (!isComplete) {
      isOnboardingRequired = true;
      targetOnboardingPath = '/customer/choose-interests';
    }
  } else if (role === 'admin') {
    targetDashboardPath = '/admin/dashboard';
  }

  const redirectTo = isOnboardingRequired ? targetOnboardingPath : (targetDashboardPath || '/customer/home');

  return {
    user,
    activeRole: role,
    isOnboardingRequired,
    targetOnboardingPath,
    targetDashboardPath,
    redirectTo,
  };
};

const updateProfile = async (userId, updates) => {
  const allowed = [
    'name', 'email', 'phone', 'gender', 'dob', 'occupation', 'profession', 'language',
    'profile_pic', 'avatarUrl', 'city', 'location', 'customerProfile',
    'vendorProfile', 'creatorProfile'
  ];

  const clean = {};
  for (const key of allowed) {
    if (updates[key] !== undefined) {
      clean[key] = updates[key];
    }
  }
  if (clean.phone) {
    const digits = String(clean.phone).replace(/\D/g, '');
    if (digits.length === 10) {
      clean.phone = `+91${digits}`;
    } else if (digits.length === 12 && digits.startsWith('91')) {
      clean.phone = `+${digits}`;
    }
  }
  if (Object.keys(clean).length === 0) throw ApiError.badRequest('No updatable fields');

  const user = await User.findByIdAndUpdate(userId, { $set: clean }, { returnDocument: 'after' });
  if (!user) throw ApiError.notFound('User not found');
  try {
    const cache = require('../utils/cache');
    await cache.deleteCache(`user:auth:${userId}`);
  } catch (err) {}
  return user;
};

const serialize = (user) => {
  if (!user) return null;
  const obj = user.toObject ? user.toObject() : { ...user };
  delete obj.password;
  delete obj.resetPasswordOtpHash;
  delete obj.resetPasswordExpires;

  if (!Array.isArray(obj.roles)) {
    obj.roles = ['customer'];
  }
  const vp = obj.vendorProfile;
  if (vp && (vp.shopName || vp.businessName || vp.businessType || (vp.categories && vp.categories.length > 0)) && !obj.roles.includes('vendor')) {
    obj.roles.push('vendor');
  }
  const cp = obj.creatorProfile;
  if (cp && (cp.displayName || cp.name || cp.handle || (cp.categories && cp.categories.length > 0)) && !obj.roles.includes('creator')) {
    obj.roles.push('creator');
  }

  return obj;
};

module.exports = { getUserById, addRole, switchRole, updateProfile, serialize };
